src/impl/realm_coordinator.cpp

```diff
@@ -882,19 +890,25 @@ void RealmCoordinator::process_available_async(Realm& realm)
 
     bool in_read = realm.is_in_read_transaction();
     auto& sg = Realm::Internal::get_shared_group(realm);
-    if (!sg) // i.e. the Realm was closed in a callback above
-        return;
     auto version = sg->get_version_of_current_transaction();
     auto package = [&](auto& notifier) {
         return !(notifier->has_run() && (!in_read || notifier->version() == version) && notifier->package_for_delivery());
     };
     notifiers.erase(std::remove_if(begin(notifiers), end(notifiers), package), end(notifiers));
+    if (notifiers.empty())
+        return;
     lock.unlock();
 
     // no before advance because the Realm is already at the given version,
     // because we're either sending initial notifications or the write was
     // done on this Realm instance
 
+    if (realm.m_binding_context) {
+        realm.m_binding_context->will_send_notifications();
```

Comment:
Isn't it wrong to notify here if we then return immediately after and not send notifications after all? If I remember correctly there is also a `did_send_notifications()` which should always be paired.

Reply:
We only return immediately after this if the `will_send_notifications()` callback closes the Realm, and in that case the previous code also didn't call `did_send_notifications()`. That might be a bug that's now just more obvious, but I don't know what it'd actually do in did_send if the Realm is already closed.